### PR TITLE
Add initial testing

### PR DESCRIPTION
**Pull request by Tankenstein**
Adds:
- Understanding permissions in token
- Understanding test types
- Dynamic jsonschema driven form for adding tests
- Test adding and confirmation flow
- Test listing

Out of scope for this pr:
- Drilldown into a single test

Still left to do with this code in later PRs:
- A lot of refactoring - this is prototype code
- More robust handling of json schema
- Missing design (for listing, improvements to be done on json schema)

### CODE REVIEW

File: src/api.ts [hunk 1]
```diff
@@ -142,3 +142,90 @@ export async function createAccessPass(
   );
   return response.data;
 }
+
+export type FieldValue = boolean | string | number;
+
+export interface TestResults {
+  details: { [key: string]: FieldValue }; // TODO: proper typing based on test type
```
Comment (StevePole) [inline src/api.ts:149]: Because json schema can support any type, any sort of strict typing is a bit futile...

File: src/api.ts [hunk 2]
```diff
@@ -142,3 +142,90 @@ export async function createAccessPass(
...
+  title: string;
+  description?: string;
+}
+
+// TODO: we are currently hardcoding only simple object -> property json schemas.
+// Should support larger range in the future.
+export interface ObjectSchema {
+  type: 'object';
+  title: string;
+  properties: { [key: string]: FieldSchema };
```
Comment (StevePole) [inline src/api.ts:192]: Could support description here as well, might provide the ability to give a bit more context to the result format.

File: src/api.ts [hunk 3]
```diff
@@ -142,3 +142,90 @@ export async function createAccessPass(
...
+}
+
+export async function fetchTestTypes(options: AuthenticatedHttpOptions): Promise<TestType[]> {
+  const response = await authenticated(options.token).get(`/api/v1/test-types`, {
+    cancelToken: options.cancelToken,
+  });
+  return response.data;
+}
+
+export type FilledSchema = { [key: string]: boolean | string | number };
```
Comment (StevePole) [inline src/api.ts:209]: We have three variations on this theme.  Should it just be one or two?

File: src/identity/IdentityPage.test.tsx [hunk 4]
```diff
@@ -71,25 +86,34 @@ describe('Identity page', () => {
...
+      await wait(() => expect(screen.queryByText(/no tests yet/i)).toBeTruthy());
       expect(screen.queryByText(/Mock QRCode: mock-sharing-code/i)).toBeFalsy();
       fireEvent.click(screen.getByText(/profile/i));
       await wait(() => expect(screen.queryByText(/Mock QRCode: mock-sharing-code/i)).toBeTruthy());
-      expect(screen.queryByText(/tests will be shown here/i)).toBeFalsy();
+      expect(screen.queryByText(/no tests yet/i)).toBeFalsy();
+    });
+
+    it('shows all your tests on the tests tab', async () => {
+      await wait(() => expect(screen.queryByText(/first middle last/i)).toBeTruthy());
```
Comment (StevePole) [inline src/identity/IdentityPage.test.tsx:99]: first middle last?  Seems an odd expectation for this test?

Comment (Tankenstein) [reply src/identity/IdentityPage.test.tsx:99]: Honestly this is just waiting for the page to load.

File: src/identity/ViewingOtherProfileHeader.tsx [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+const LinkButton = Button as React.FC<ButtonProps & LinkProps>;
+// TODO: the styling on this header is very ad-hoc. Let's move it into the theme.
+// TODO: this maybe doesn't belong in this module. Where does it belong?
+export const ViewingOtherProfileHeader = () => {
+  const { userId } = useAuthentication();
+  return (
+    <Box sx={{ backgroundColor: 'primary', color: 'background' }}>
+      <Flex px={3} py={2} sx={{ justifyContent: 'space-between', maxWidth: '600px' }} mx="auto">
+        <Heading as="h2" sx={{ lineHeight: 1.8 }}>
+          Patient profile
```
Comment (StevePole) [inline src/identity/ViewingOtherProfileHeader.tsx:15]: Not sure about this language.  I know we've been using it, but would be weird in the passport use case.  Wondering if tt could be more like "Viewing: Steve Pole".

File: src/testing/AddTestForm.tsx [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  const form = useFormik({
+    initialValues: getInitialValues(testType.resultsSchema),
+    onSubmit(details: FilledSchema) {
+      onComplete({ testTypeId: testType.id, results: { details } });
+    },
+  });
+
+  return (
+    <>
+      <Text>Result</Text>
```
Comment (StevePole) [inline src/testing/AddTestForm.tsx:27]: You could use the schema title here if there's one present.

File: src/testing/AddTestForm.tsx [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+export const AddTestForm = ({
+  testType,
+  onComplete,
+}: {
+  testType: TestType;
+  onComplete: (command: CreateTestCommand) => any;
+}) => {
+  // TODO: validation
+  // TODO: supporting more arbitrary schemas
```
Comment (StevePole) [inline src/testing/AddTestForm.tsx:17]: Main thing I would add is enums (selects/radios). Don't think we need to get into nested object, array, or oneOf/allOf.

File: src/testing/AddTestPage.test.tsx [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+        </Route>
+      </Router>,
+    ).debug;
+  });
+
+  describe('when testing yourself', () => {
+    it('lets you add a test result', async () => {
+      history.push('/users/mock-user/add-test');
+      await wait(() => expect(screen.queryByText(/test type/i)).toBeTruthy());
+      expect(createTestMock).not.toHaveBeenCalled();
```
Comment (StevePole) [inline src/testing/AddTestPage.test.tsx:57]: Would be tempted to move the above lines into their own expectation, just as this test is massive and they don't directly relate to the described situation.

File: src/identity/ViewingOtherProfileHeader.tsx (hunk 5 above)
Comment (Tankenstein) [reply src/identity/ViewingOtherProfileHeader.tsx:15]: In a lot of screens we show the name anyhow. Anyhow, this is feedback for Duncan.

File: src/testing/AddTestForm.tsx (hunk 6 above)
Comment (Tankenstein) [reply src/testing/AddTestForm.tsx:27]: It looks a bit ridiculous - the select is on the top with the title, under which you have the same exact title

**Pull request merged**